### ISSUE TITLE
SendGrid client has a memory leak

### ISSUE THREAD

**Issue by kirk-quinbar** [sendgrid/sendgrid-csharp#658]
#### Issue Summary

Over time the sendgrid client calling the mail api seems to be holding on to or leaking memory. I reported a similar error in 2016 but i thought it was fixed which is why we were looking at this again. If you have an application where you are sending out hundreds or thousands of emails, you can see memory usage build up. We actually did a larger test of sending 20,000 emails on one of our servers, and the memory leak ate up half of available memory. whatever it is in the client that is holding on to it, never released the memory even days later and we had to reboot the server to clear it.

#### Steps to Reproduce

1. open up the attached sample project
2. add appropriate apikey, username, and password constants
3. there is a TestWithApi and TestWithSmtp method for your comparison
4. add appropriate from and to addresses to both test methods
5. run each one separately while viewing the diagnostics window (assuming visual studio 2017)
6. the testwithapi will continue to increase in memory usage and never release anything
7. the testwithsmtp will get up to a certain memory usage (like 20mb or less) and then stay level.

It looks to me like its related to the _sslstream and/or the Connection object

#### Technical details:

* sendgrid-csharp Version: master (9.9.0)
* Platform (.NET)
* .NET Version: 4.6
![memorysnapshot4](https://user-images.githubusercontent.com/30601457/35169684-113b753e-fd23-11e7-9070-fe7167fe94e2.png)
![memoryusage](https://user-images.githubusercontent.com/30601457/35169685-11462574-fd23-11e7-918a-8c319a15fb9b.png)
![memorysnapshot1](https://user-images.githubusercontent.com/30601457/35169686-1150baa2-fd23-11e7-9234-d62e7c40893c.png)
![memorysnapshot2](https://user-images.githubusercontent.com/30601457/35169687-11634a32-fd23-11e7-916c-c327e2de53c3.png)
![memorysnapshot3](https://user-images.githubusercontent.com/30601457/35169688-11727476-fd23-11e7-9ab4-d4503293a1d5.png)
[sendgridemailapi.zip](https://github.com/sendgrid/sendgrid-csharp/files/1647816/sendgridemailapi.zip)

.2

**Comment by thinkingserious** [comment]
Hi @kirk-quinbar,

Thanks for the detailed feedback, very useful!

I'm guessing there may be an issue with how we are using the [HTTPClient](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/). I've added this to our backlog for further investigation.

Thanks again for putting this on the radar!

With Best Regards,

Elmer

**Comment by Jericho** [comment]
@kirk-quinbar, I ran your sample project and I also observed that memory utilization keeps going up and is not released, very similar to what you observed. However, I think the issue is not so much a bug and/or memory leak in sendgrid-csharp but rather a flaw in your testing. Let me explain.

SendGrid's csharp client is dependent on Microsoft's HttpClient to handle the actual HTTP communication between your workstation and their API. This http client implements the IDisposable interface which leads a lot of developer to assume that it's intended to be disposed of quickly between each http request while in fact that's not the case. There are countless blog posts (such as this [one](https://blogs.msdn.microsoft.com/alazarev/2017/12/29/disposable-finalizers-and-httpclient/)) and questions on StackOverflow (such as [this one](https://stackoverflow.com/questions/15705092/do-httpclient-and-httpclienthandler-have-to-be-disposed) for example) that debate whether you should dispose the http client between each request of if you should reuse it and the consensus is that you should hold on to your http client as long as possible. Simon Timms wrote, in my opinion, [the best article on this topic](https://blogs.msdn.microsoft.com/alazarev/2017/12/29/disposable-finalizers-and-httpclient/).

Your testing project instantiate a new `SendGridClient` for each request which in turn creates a new http client in its constructor. The net effect is that you are instantiating a new http client for every single http request and this is the root cause of the high memory utilization in your testing project.

By the way, this situation with the http client has caused so many problem and so much confusion that Microsoft has decided to address it in the [upcoming .NET Core 2.1](https://blogs.msdn.microsoft.com/dotnet/2018/02/02/net-core-2-1-roadmap/) with a `HttpClientFactory` class that will manage a "pool" of http client in memory and will ensure they are reused when appropriate and released after a reasonable timeout). Here's an article about the [reasoning behind this new factory class](https://www.stevejgordon.co.uk/introduction-to-httpclientfactory-aspnetcore).

Long story short: you can improve your testing project by instantiating a single HttpClient and reusing it for each of your HTTP requests. I modified your project and I observed much better memory utilization.

__Modified `Program.cs` that reuses the same HTTP client for each request:__
[Program.cs.txt](https://github.com/sendgrid/sendgrid-csharp/files/1692342/Program.cs.txt)

__Memory utilization after modification:__
![sshot-11](https://user-images.githubusercontent.com/112710/35773256-6d6a553e-091a-11e8-8fec-080f6b06d765.png)

**Comment by thinkingserious** [comment]
Thank you for the excellent explanation @Jericho!

**Comment by vlmironoff** [comment]
This is the information I was looking for.
Couple questions:
1. Would you create a constructor overload for SendgridClient that will take the HttpClient as a parameter and use it? Then it will be on SDK consumer to set/maintain an injected HttpClient's lifetime.

2. Meanwhile, is it safe to register SendgridClient as Singleton in a DI container to handle email traffic from different threads?

Thank you.

**Comment by kirk-quinbar** [comment]
1. yes that is one way to fix the issue.
2. in my scenario, i needed to pass in unknown apikeys from a database to
send emails, so i ended up making a static concurrentdictionary of sendgrid
clients and then just reusing one based on a keyvalue. i suppose i might
have been able to just make a single static http client but just went the
route of the sendgrid client. each client uses up a small amount of memory
but there is a finite number of apikeys in our database so for now we are
just watching the memory usage. it seemed to be pretty stable since we are
reusing cilents.

Kirk

**Comment by TheGreatYamcha** [comment]
I believe SendGrid api should still give the dispose option. Since the client isn't thread safe, it's not clear that we should reuse clients (through a pool) in a multithreaded environment.

**Comment by CarlLeth** [comment]
@Jericho If SendGridClient instantiates an IDisposable resource, then it should also be IDisposable. Not doing so is introducing a memory leak. So the OP is correct: SendGridClient has a memory leak and should implement IDisposable to fix it.

**Comment by Jericho** [comment]
@CarlLeth I realize that HttpClient implements the IDisposable interface which leads a reasonable person to assume that every instance should be disposed quickly but I hope you read the articles I referenced because they discuss at length why Microsoft's HttpClient is not intended to be disposed after each and every request __despite the fact that it implements the `IDisposable` interface__. 

It's been 2 years since the OP raised this issue and I responded to him so please forgive me if the details are not fresh in my memory but, if I remember correctly, the biggest issue is that you will exhaust the available connections on your machine if you instantiate and dispose too many HTTP clients too quickly which is what the OP was doing and lead me to recommend that he should reuse the client.

Quoting from [Simon Timm's excellent blog post](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/):
> For years we have been conditioned to dispose of objects that implement IDisposable and many refactoring tools like R# and CodeRush actually warn if you don’t. In this case disposing of HttpClient was the wrong thing to do. It is unfortunate that HttpClient implements IDisposable and encourages the wrong behaviour.
